File: ChoiSubin/프로그래머스/카카오/합승_택시_요금.cpp
#include <string>
#include <vector>
#include <cstring>
#define INF 987654321
using namespace std;
int arr[201][201];
int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = 0;

    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            if (i == j)
                arr[i][j] = 0;
            else
                arr[i][j] = INF;
        }
    }

    for (auto f : fares)
    {
        arr[f[0]][f[1]] = f[2];
        arr[f[1]][f[0]] = f[2];
    }

    //플로이드 와샬로 모든 정점에서 부터 최단거리 계산
    for (int k = 1; k <= n; k++)
    {
        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                if (i != j && i != k && k != j && arr[i][k]!=INF && arr[k][j]!=INF)
                {
                    if (arr[i][k] + arr[k][j] < arr[i][j])
                    {
                        arr[i][j] = arr[i][k] + arr[k][j];
                        arr[j][i] = arr[i][j];
                    }
                }
            }
        }
    }

    answer = arr[s][a] + arr[s][b]; //둘이 따로 탔을때 드는 비용
    for (int i = 1; i <= n; i++) //모든 정점 돌면서 둘이 합승했을때 최솟값 있으면 answer에 넣어주기
    {
        if (arr[s][i] != INF && arr[i][a] != INF && arr[i][b] != INF) //i정점을 들렸을때 도착할 수 있으면
        {
            answer = (answer > arr[s][i] + arr[i][a] + arr[i][b]) ? arr[s][i] + arr[i][a] + arr[i][b] : answer;
            //만약 중간에 i정점을 들려서 도착하는 비용이 더 싸면 그 비용을 대입해주고 아니면 그냥 원래 값 넣어주기
        }
    }
    return answer;
}

int main()
{
    solution(6, 4, 5, 6, {{2, 6, 6}, {6, 3, 7}, {4, 6, 7}, {6, 5, 11}, {2, 5, 12}, {5, 3, 20}, {2, 4, 8}, {4, 3, 9}} );
    return 0;
}

File: Se0nAh/programmers/합승_택시_요금.js
// 1시간 풀고도 틀림 ㅎㅎ; 다시 수정...
// 정확성: 40.0
// 효율성: 36.7
// 합계: 76.7 / 100.0
function solution(n, s, a, b, fares) {
    const INF = 100001
    var answer = INF;
    
    var arr, i, j, initialMat, lowestCostMat = [];
    for (i = 0; i < n+1; i += 1) {
        arr = [];
        for (j = 0; j < n+1; j += 1) {
            if (i === j) arr[j] = 0
            else arr[j] = INF;
        }
        lowestCostMat[i] = arr;
    }
    
    fares.forEach(fare => {
        lowestCostMat[fare[0]][fare[1]] = fare[2]
        lowestCostMat[fare[1]][fare[0]] = fare[2]
    }) 


    function floyedWarshall (g){
        for (let t = 1; t < n+1 ; t+=1){ // 경유 노드
            for (let s = 1; s < n+1; s+=1){ // 시작 노드
                if (s === t) continue
                for (let a =1; a < n+1; a+=1) { // 도착 노드
                    if (s === a) continue
                    if (g[s][a] > g[s][t] + g[t][a]){
                        g[s][a] = g[s][t] + g[t][a]
                    }
                }
            }
        }
    }
    floyedWarshall(lowestCostMat)
    
    
    console.log(lowestCostMat)
        for (let t = 1; t < n + 1 ; t++) {            
            // 각 지점에서 a 와 b 목적지까지 가는 데 드는 최소 비용 계산
            // s에서 각지점까지 가는 데 드는 최소  비용 계산\
            // 두값을 더해 나온 값들 중 최소값??
            answer = Math.min(answer, lowestCostMat[t][a] + lowestCostMat[t][b] + lowestCostMat[s][t])
            console.log(answer)
        }
    
    return answer;
}